### PR TITLE
[Collections] Allow custom UIEdgeInsets

### LINKED ISSUE #1507: [Collections] Have a way to change Cards style margins
### Overview
The Cards style has a fixed right/left margin of 8pt. On large device, if the collection is full width, the cards' width is too big.
It should have a way to adjust the margin or the width of the cells.

### PR DESCRIPTION
**Pull request by romoore**
MDCCollectionViewController implements custom logic for computing the
`UIEdgeInsets` depending on the `MDCCollectionViewCellStyle` and the
`MDCCollectionViewCellLayoutType`. Subclasses need to override this
functionality to customize their presentation.

* MDCCollectionViewController will use the public
  `-collectionView:layout:section` to compute the width of its
   supplementary view instead of the internal implementation directly.
* Update the Collections layout example to have different horizontal
  edge insets depending on the view's horizontal size class.

Closes #1507

### CODE REVIEW

File: components/CollectionCells/examples/CollectionCellsLayoutExample.m
Comment (gambard) [inline components/CollectionCells/examples/CollectionCellsLayoutExample.m]: Don't you prefer to add the example in the Collections category?

Comment (romoore) [reply components/CollectionCells/examples/CollectionCellsLayoutExample.m]: Yep, good callpout. I'm working on writing a new "Collections" example and an Earl Grey test to verify it.

File: components/Collections/examples/CollectionsSectionInsetsExample.m
Comment (willlarche) [inline components/Collections/examples/CollectionsSectionInsetsExample.m]: Usually we use @propertys in an objc extension and let the ivars be free. If you're overriding both the setter and the getter you'll _also_ have to declare an ivar but that's a super exception.

File: components/Collections/examples/CollectionsSectionInsetsExample.m
Comment (willlarche) [inline components/Collections/examples/CollectionsSectionInsetsExample.m]: Move the CBC methods to the bottom of the file (or into a -supplemental) in an extension "(CatalogByConvention)"

File: components/Collections/examples/CollectionsSectionInsetsExample.m
Comment (willlarche) [inline components/Collections/examples/CollectionsSectionInsetsExample.m]: catalogDescription is a description of the component itself. So, you can delete this since there's one in CollectionsSimpleDemo.m.

File: components/Collections/examples/CollectionsSectionInsetsExample.m
Comment (willlarche) [inline components/Collections/examples/CollectionsSectionInsetsExample.m]: Instead of creating a local variable, cast the sender in the method signature by changing 'id' to 'UISwitch*'.

File: components/Collections/examples/CollectionsSectionInsetsExample.m
Comment (willlarche) [inline components/Collections/examples/CollectionsSectionInsetsExample.m]: This would be best in a -Supplemental file.

File: components/Collections/examples/CollectionsSectionInsetsExample.m
Comment (willlarche) [inline components/Collections/examples/CollectionsSectionInsetsExample.m]: This would be best in a -Supplemental file.

File: components/Collections/examples/CollectionsSectionInsetsExample.m
Comment (willlarche) [inline components/Collections/examples/CollectionsSectionInsetsExample.m]: This whole section would be best in a -Supplemental file since the 'meat' of the instruction here is the UICollectionViewDelegateFlowLayout changes.

File: components/Collections/examples/CollectionsSectionInsetsExample.m
Comment (willlarche) [inline components/Collections/examples/CollectionsSectionInsetsExample.m]: Usually we use @propertys where we get the ivar for free. The exception being when we override the setter and the getter of the property when we need to have both the @property declaration and the explicit ivar.

File: components/Collections/examples/CollectionsSectionInsetsExample.m
Comment (willlarche) [inline components/Collections/examples/CollectionsSectionInsetsExample.m]: catalogDescription is a description of the component itself which has already been done in CollectionsSimpleDemo.m. So, you can delete this.

File: components/Collections/examples/CollectionsSectionInsetsExample.m
Comment (willlarche) [inline components/Collections/examples/CollectionsSectionInsetsExample.m]: Move the catalog by convention methods to an extension called (CatalogByConvention) and put it either at the end of the file or in a -Supplemental.

File: components/Collections/examples/CollectionsSectionInsetsExample.m
Comment (willlarche) [inline components/Collections/examples/CollectionsSectionInsetsExample.m]: This would be a good candidate to move to a -Supplemental file.

File: components/Collections/examples/CollectionsSectionInsetsExample.m
Comment (willlarche) [inline components/Collections/examples/CollectionsSectionInsetsExample.m]: This section also would be a good candidate to move to the -Supplemental file.

File: components/Collections/examples/CollectionsSectionInsetsExample.m
Comment (willlarche) [inline components/Collections/examples/CollectionsSectionInsetsExample.m]: This entire section would be a good candidate to move to the -Supplemental file since the 'meat' of the instruction is the UICollectionViewDelegateFlowLayout section.

File: components/Collections/examples/CollectionsSectionInsetsExample.m
Comment (willlarche) [inline components/Collections/examples/CollectionsSectionInsetsExample.m]: You can get rid of the local variable by casting sender in the method signature by changing id to UISwitch *

File: components/Collections/tests/earlgrey/MDCCollectionViewControllerTests.m
Comment (willlarche) [inline components/Collections/tests/earlgrey/MDCCollectionViewControllerTests.m]: I think @"collectionView" is good candidate for a const string now.

File: components/Collections/tests/earlgrey/MDCCollectionViewControllerTests.m
Comment (willlarche) [inline components/Collections/tests/earlgrey/MDCCollectionViewControllerTests.m]: @"back_bar_button" would also be a good candidate for a const string.

File: components/Collections/tests/earlgrey/MDCCollectionViewControllerTests.m
Comment (willlarche) [inline components/Collections/tests/earlgrey/MDCCollectionViewControllerTests.m]: If element will always be a UICollectionView *, then cast it right in the block declaration.

File: components/Collections/tests/earlgrey/MDCCollectionViewControllerTests.m
Comment (willlarche) [inline components/Collections/tests/earlgrey/MDCCollectionViewControllerTests.m]: Same note about casting in the block declaration instead of using a local variable.

File: components/Collections/tests/earlgrey/MDCCollectionViewControllerTests.m
Comment (willlarche) [inline components/Collections/tests/earlgrey/MDCCollectionViewControllerTests.m]: Since we can use this for all the tests, move it to a more general file.

**Review by willlarche — CHANGES_REQUESTED** [review]
All style notes. Content is solid!

File: components/Collections/examples/CollectionsSectionInsetsExample.m
Comment (romoore) [reply components/Collections/examples/CollectionsSectionInsetsExample.m]: Done

File: components/Collections/tests/earlgrey/MDCCollectionViewControllerTests.m
Comment (romoore) [reply components/Collections/tests/earlgrey/MDCCollectionViewControllerTests.m]: It doesn't seem to accept that as a signature, and the explicit cast makes maintainers recognize just how brittle it is.

File: components/Collections/examples/CollectionsSectionInsetsExample.m
Comment (romoore) [reply components/Collections/examples/CollectionsSectionInsetsExample.m]: Done

File: components/Collections/examples/CollectionsSectionInsetsExample.m
Comment (romoore) [reply components/Collections/examples/CollectionsSectionInsetsExample.m]: Done

File: components/Collections/examples/CollectionsSectionInsetsExample.m
Comment (romoore) [reply components/Collections/examples/CollectionsSectionInsetsExample.m]: Done

File: components/Collections/examples/CollectionsSectionInsetsExample.m
Comment (romoore) [reply components/Collections/examples/CollectionsSectionInsetsExample.m]: Done

File: components/Collections/examples/CollectionsSectionInsetsExample.m
Comment (romoore) [reply components/Collections/examples/CollectionsSectionInsetsExample.m]: Done

File: components/Collections/examples/CollectionsSectionInsetsExample.m
Comment (romoore) [reply components/Collections/examples/CollectionsSectionInsetsExample.m]: Done

File: components/Collections/tests/earlgrey/MDCCollectionViewControllerTests.m
Comment (romoore) [reply components/Collections/tests/earlgrey/MDCCollectionViewControllerTests.m]: Doesn't seem to play well with the signature

Comment (willlarche) [reply components/Collections/tests/earlgrey/MDCCollectionViewControllerTests.m]: Ack!

**Review by willlarche — APPROVED** [review]

File: components/Collections/tests/earlgrey/MDCCollectionViewControllerTests.m
Comment (romoore) [reply components/Collections/tests/earlgrey/MDCCollectionViewControllerTests.m]: I was wrong, and when I tried again it was fine. Switched.

File: catalog/MDCCatalog/MDCNodeListViewController.swift
Comment (romoore) [inline catalog/MDCCatalog/MDCNodeListViewController.swift]: Remove parentheses

File: components/Collections/tests/earlgrey/MDCCollectionViewControllerTests.m
Comment (romoore) [inline components/Collections/tests/earlgrey/MDCCollectionViewControllerTests.m:116]: Add pragma mark

**Review by ajsecord — APPROVED** [review]

**Comment by ianegordon** [comment]
I've asked Felix to take a look as he recently dealt with EdgeInsets.

**Review by ianegordon — CHANGES_REQUESTED** [review]
Waiting for a LGTM from Felix.

**Comment by ianegordon** [comment]
@felixaemiliano will be reviewing / testing this on his project tomorrow.

File: components/Collections/src/MDCCollectionViewFlowLayout.m
Comment (felixaemiliano) [inline components/Collections/src/MDCCollectionViewFlowLayout.m:394]: Is the intention that the default insets be derived from -insetsAtSectionIndex? Currently this is not the case: the insets are only set to that value if the collectionView's delegate is nil (which is unlikely).

Perhaps the correct implementation would be to set the insets here to [self insetsAtSectionIndex] and conditionally override below if the flowLayoutDelegate responds to the selector which allows subclasses to customize the insets.

**Review by felixaemiliano — COMMENTED** [review]
This is definitely a welcome change that allows subclasses more control, thanks for adding it!

However, Classroom overrides the -cellWidthAtSectionIndex: method, which means we will not be affected by this change.

File: components/Collections/src/MDCCollectionViewFlowLayout.m
Comment (romoore) [reply components/Collections/src/MDCCollectionViewFlowLayout.m:394]: Good catch, I can switch the implementation to start with the internals and then call the delegate.

**Review by ianegordon — APPROVED** [review]
LGTM.

**Pull request merged**
